Use SystemScalarConverter for Diagram transmogrification

jwnimmer-tri opened the pull request:
There are three major features happening here:
1. Diagram's set of supported scalar types will come from (i) the list that SystemScalarConverter curates, but then be limited by (ii) the runtime support for transmogrification provided by whatever subtypes are added into the Diagram.  This allows us to change the set of default scalar types by editing one place (SystemScalarConverter) instead of also having to patch Diagram.
2. Diagram gains support for preserving the Diagram subtype during transmogrification.
3. The NVI methods DoToFoo in System go away, again so that SystemScalarConverter owns more of the responsibility, and thus we have less repetition.

I think it makes sense to consider all of these things together as a whole, but please let me know if separate PRs would be preferable.  The commits are curated, so can still be browsed or reviewed one at a time if desired.

Commits:
- Add SystemImpl::get_mutable_system_scalar_converter.
  - This is not unit tested directly (due to attorney-client privilege), but the diagram_test cases cover it in the next commit.
- Port Diagram to SystemScalarConverter.
- Add Doxygen overview for Diagram transmogrification.
- Remove System::DoTo transmog NVI hooks.

Relates #7039.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7091)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@amcastro-tri for feature review, please.
+@soonho-tri for platform review per rotation, please.
/CC @sherm1 in case you want to tag along for a design review.

amcastro-tri commented:
future review pass done. I think mostly minor comments. Once resolved <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 662 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyVTtVJ3jzzPHMc17_:-KuyVTtVJ3jzzPHMc17a:bv6ec0k) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/diagram.h#L662)):*
> ```objc
>   template <typename U>
>   explicit Diagram(const Diagram<U>& other)
>       : Diagram(other.template ConvertScalarType<T>()) {}
> ```

per style guide this should probably be moved closed to the beginning of the `public:` section together with the other constructors.

---

*[drake/systems/framework/diagram.h, line 1181 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyQkWidm7-g2eNffqV:-KuyQkWidm7-g2eNffqW:b-f7s5ss) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/diagram.h#L1181)):*
> ```objc
> 
>       // Match the result's name to its originator.
>       new_system->set_name(old_system->get_name());
> ```

Shouldn't the `System<T>` constructor be doing this?

---

*[drake/systems/framework/system.h, line 1057 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuyds2joyAA_a4HUWXf:-Kuyds2joyAA_a4HUWXg:bum3g2u) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system.h#L1057)):*
> ```objc
>     if (result) {
>       // Match the result's name to its originator.
>       result->set_name(this->get_name());
> ```

wouldn't it be easier/safer to move this to one of the `System`'s constructors?

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 278 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyaCmo6xjHnWrobE1G:-KuyaCmo6xjHnWrobE1H:bm8sxwm) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L278)):*
> ```objc
> @code
> namespace sample {
> class MyDiagram : public Diagram<double> {
> ```

should this be marked `final`?

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 296 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuy_CAWhno3xmyJpnvk:-Kuy_CAWhno3xmyJpnvl:bajjwii) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L296)):*
> ```objc
> By default, even subclasses of a `Diagram<U>` will convert to a `Diagram<T>`,
> discarding the diagram subclass details.  For example, in the above sample
> code, `MyDiagram<double>::ToAutoDiffXd()` will return a object of runtime type
> ```

The only problem I see with this text is that `MyDiagram` in the example provided is not a template class (which is good to highlight the point of the class not necessarily needing to be templated on a T). Maybe fix to use `OtherDiagram<T>` as example?

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 306 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyaFakO-3lPlD_Hrwn:-KuyaFakO-3lPlD_Hrwo:bm8sxwm) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L306)):*
> ```objc
> namespace sample {
> template <typename T>
> class MyDiagram<T> : public Diagram<T> {
> ```

should this be marked `final`?

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuy_nHaFjIB2gVrI_0-:-Kuy_nHaFjIB2gVrI_00:bb9nytn) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L309)):*
> ```objc
>  public:
>   MyDiagram() : Diagram<T>(SystemTypeTag<sample::MyDiagram>{}) {
>     // ... same body as prior example (only the above line changes) ...
> ```

BTW, well.. not really, this one IS templated on T now.

---

*[drake/systems/framework/test/diagram_test.cc, line 409 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyWVLQiv4sxE_K8ZIE:-KuyWVLQiv4sxE_K8ZIF:b-2t2qb6) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L409)):*
> ```cpp
>   // we cannot transmogrify the Diagram to AutoDiffXd.
>   auto diagram_with_abstract = std::make_unique<ExampleDiagram>(kSize, true);
>   EXPECT_THROW(diagram_with_abstract->ToAutoDiffXd(), std::exception);
> ```

I believe this is true now because `ConstantValueSource` does not provide the scalar type conversion constructor. But I believe also we will want to provide that scalar conversion constructor soon. Wouldn't that break this test? 
I'd maybe write this test in terms of a locally defined class that doesn't provide scalar conversion. Maybe worth also placing this test in its own `TEST_F()`.

---

*[drake/systems/framework/test/diagram_test.cc, line 415 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyYsdMLtaAr5bYG9GU:-KuyYsdMLtaAr5bYG9GV:b9ytwdv) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L415)):*
> ```cpp
> TEST_F(DiagramTest, ToSymbolic) {
>   // Upcast to `Diagram&` so that is_symbolic_convertible asserts the
>   // transmogrification result is merely a Diagram, not an ExampleDiagram.
> ```

I don't quite understand this comment. What I do understand is that `is_symbolic_convertible()` expects a template class argument. Is that why you have the upcast to `Diagram`? or am I missing something?

---

*[drake/systems/framework/test/diagram_test.cc, line 420 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyY6aHJDlUFnfTZczq:-KuyY6aHJDlUFnfTZczr:bmfwjue) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L420)):*
> ```cpp
> 
>   // No symbolic support when one of the subsystems does not declare support.
>   auto diagram_with_abstract = std::make_unique<ExampleDiagram>(kSize, true);
> ```

same comment regarding `ConstantValueSource` not being convertible now but most likely convertible in the near future.

---

*[drake/systems/framework/test/diagram_test.cc, line 703 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuyadym1wdxXvS_QUBw:-Kuyadym1wdxXvS_QUBx:b-qfjhxy) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L703)):*
> ```cpp
> class FeedbackDiagram : public Diagram<T> {
>  public:
>   FeedbackDiagram() : Diagram<T>(SystemTypeTag<systems::FeedbackDiagram>{}) {
> ```

BTW maybe add comment here explaining why use the more unusual way of supporting scalar type conversion in this particular example.

---

*[drake/systems/framework/test/diagram_test.cc, line 758 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuyb83hODDZOz9Ka6I3:-Kuyb83hODDZOz9Ka6I4:b-fdmdbe) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L758)):*
> ```cpp
>   const FeedbackDiagram<double> dut;
>   EXPECT_TRUE(is_autodiffxd_convertible(dut, [](const auto& converted) {
>     EXPECT_FALSE(converted.HasAnyDirectFeedthrough());
> ```

BTW, this is pretty cool!

---

*[drake/systems/framework/test/diagram_test.cc, line 766 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuybJnXY_DviceOLb_f:-KuybJnXY_DviceOLb_g:b-4v4hbk) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L766)):*
> ```cpp
>   // Diagram subclasses that declare a specific SystemTypeTag but then use a
>   // subclass at runtime will fail-fast.
>   class SubclassOfFeedbackDiagram : public FeedbackDiagram<double> {};
> ```

BTW, nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-:-KuyeIh1dRHAcYHLNqox:brkrgcg)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Excellent, thanks.  Checkpointing my fixes for now.

---

Review status: 4 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 662 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyVTtVJ3jzzPHMc17_:-Kuz0orxkR9Xkzbgp7_u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/diagram.h#L662)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

per style guide this should probably be moved closed to the beginning of the `public:` section together with the other constructors.
</blockquote></details>

Done.

---

*[drake/systems/framework/diagram.h, line 1181 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyQkWidm7-g2eNffqV:-Kuz1XiMDv3bhxEFUoVw:bqc5a7j) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/diagram.h#L1181)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Shouldn't the `System<T>` constructor be doing this?
</blockquote></details>

Yeah, this is been brittle for a while and was on the plan to fix.  But now I've filed to #7108 to make sure we don't forget.

---

*[drake/systems/framework/system.h, line 1057 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuyds2joyAA_a4HUWXf:-Kuz1kEsTD4qz2fA6LMO:bbpit4w) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system.h#L1057)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

wouldn't it be easier/safer to move this to one of the `System`'s constructors?
</blockquote></details>

#7108 

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 278 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyaCmo6xjHnWrobE1G:-Kuz1rw1S5ffUcGR30il:bpc3vbo) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L278)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

should this be marked `final`?
</blockquote></details>

It actually doesn't need to be, because the subclass type is being discarded during transmogrification.

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 296 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuy_CAWhno3xmyJpnvk:-Kuz3I32jJ97QB7RstvR:bmmjn4c) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L296)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

The only problem I see with this text is that `MyDiagram` in the example provided is not a template class (which is good to highlight the point of the class not necessarily needing to be templated on a T). Maybe fix to use `OtherDiagram<T>` as example?
</blockquote></details>

Done.  I removed the `<double>` here, and also renamed the second example to better distinguish it.

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 306 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyaFakO-3lPlD_Hrwn:-Kuz21fer4O_XtfA_CzH:b-l6ydw3) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L306)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

should this be marked `final`?
</blockquote></details>

Yes!  Done.

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 309 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuy_nHaFjIB2gVrI_0-:-Kuz2Ht8ea7wEgiB8iZP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/system_scalar_conversion_doxygen.h#L309)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, well.. not really, this one IS templated on T now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-:-Kuz3g5QNFycE3f3tjhr:bhfi9i2)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 1 of 2 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 1181 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyQkWidm7-g2eNffqV:-Kuz52Z1ThMkHK-suiVT:b1mtb2l) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/diagram.h#L1181)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yeah, this is been brittle for a while and was on the plan to fix.  But now I've filed to #7108 to make sure we don't forget.
</blockquote></details>

SGTM. thanks

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 296 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuz5Nzmg4AAR6HZmhmC:-Kuz5Nzmg4AAR6HZmhmD:b-apqs7s) ([raw file](https://github.com/robotlocomotion/drake/blob/b98ac0f70b682aaf42b722a26d57e9497c868376/drake/systems/framework/system_scalar_conversion_doxygen.h#L296)):*
> ```objc
> By default, even subclasses of a `Diagram<U>` will convert to a `Diagram<T>`,
> discarding the diagram subclass details.  For example, in the above sample
> code, `MyDiagram::ToAutoDiffXd()` will return a object of runtime type
> ```

BTW, typo. "an object"

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 301 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuz5zHnGz3mZ5Joj03v:-Kuz5zHnGz3mZ5Joj03w:be75knh) ([raw file](https://github.com/robotlocomotion/drake/blob/b98ac0f70b682aaf42b722a26d57e9497c868376/drake/systems/framework/system_scalar_conversion_doxygen.h#L301)):*
> ```objc
> 
> In the unusual case that the Diagram's subclass must be preserved during
> conversion, a ::drake::systems::SystemTypeTag should be used:
> ```

BTW could you comment here on when we'd like "the Diagram's subclass must be preserved during conversion"?
also, is it a defect marking a `Diagram` subclass `final` but forget to do `Diagram<T>(SystemTypeTag<sample::SpecialDiagram>{})` in the constructor?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 301 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuz5zHnGz3mZ5Joj03v:-Kuz7p3beVCRZHPANYlX:bqcpmuw) ([raw file](https://github.com/robotlocomotion/drake/blob/b98ac0f70b682aaf42b722a26d57e9497c868376/drake/systems/framework/system_scalar_conversion_doxygen.h#L301)):*
> ... when we'd like "the Diagram's subclass must be preserved during conversion"?

Something like "you'll know when you need it".  Its unusual.  Not sure what I could say.

> Is it a defect marking a Diagram subclass final but forget to do SystemTypeTag?

Nope, that would be just fine.

---

*[drake/systems/framework/test/diagram_test.cc, line 409 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyWVLQiv4sxE_K8ZIE:-Kuz7C7xUKyMtzcyU8fy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L409)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I believe this is true now because `ConstantValueSource` does not provide the scalar type conversion constructor. But I believe also we will want to provide that scalar conversion constructor soon. Wouldn't that break this test? 
I'd maybe write this test in terms of a locally defined class that doesn't provide scalar conversion. Maybe worth also placing this test in its own `TEST_F()`.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/diagram_test.cc, line 415 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyYsdMLtaAr5bYG9GU:-Kuz7D7lrBokNg8okufu:bg1pmo5) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L415)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I don't quite understand this comment. What I do understand is that `is_symbolic_convertible()` expects a template class argument. Is that why you have the upcast to `Diagram`? or am I missing something?
</blockquote></details>

Done.  Does the new way read better?

---

*[drake/systems/framework/test/diagram_test.cc, line 420 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyY6aHJDlUFnfTZczq:-Kuz4rXHz1kmRoN57vec:bs4q9op) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L420)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

same comment regarding `ConstantValueSource` not being convertible now but most likely convertible in the near future.
</blockquote></details>

Ditto.

---

*[drake/systems/framework/test/diagram_test.cc, line 703 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-Kuyadym1wdxXvS_QUBw:-Kuz4kH7F6_UZAcwbsIs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L703)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW maybe add comment here explaining why use the more unusual way of supporting scalar type conversion in this particular example.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Checkpoint (only a minor comment on comment)

---

Reviewed 4 of 6 files at r1, 1 of 2 files at r2, 1 of 2 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/diagram.h, line 678 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuzMJqHdDhLaAYLoWuv:-KuzMJqHdDhLaAYLoWuw:bgutu5n) ([raw file](https://github.com/robotlocomotion/drake/blob/032dde2fc83e5b69d664a7119f172418255cc975/drake/systems/framework/diagram.h#L678)):*
> ```objc
>   ///
>   /// Declares scalar-type conversion support using @p converter.  Support for
>   /// a given pair of types `S, T` to convert from and to will be enabled only
> ```

BTW, nit, isn't it clearer to keep using `U, T` here instead of `S, T`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-:-KuzUr0a5XmMU_drFLXE:b3uh5r5)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/test/diagram_test.cc, line 415 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuyYsdMLtaAr5bYG9GU:-KuzYzko8-RAckiU6HOl:b362gky) ([raw file](https://github.com/robotlocomotion/drake/blob/4209e571c4ea51057d0550ae647a7afe62d56b5a/drake/systems/framework/test/diagram_test.cc#L415)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Done.  Does the new way read better?
</blockquote></details>

Better, thanks!. I was wondering if I had missed something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 2 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuzbTkqPn20JCqoU-PO:-KuzbTkqPn20JCqoU-PP:bqhe6g9) ([raw file](https://github.com/robotlocomotion/drake/blob/09c52208c39c8fe7f7ba48237b0257c78b551b5b/drake/systems/framework/system_scalar_conversion_doxygen.h#L25)):*
> ```objc
> @internal
> N.B. The code sample immediately below must be kept in sync with the identical
> code in test/system_scalar_conversion_doxygen_test.cc.  This ensures that the
> ```

BTW, do we want/plan to have a copy of the newly added code in this `.h` file in `test/system_scalar_conversion_doxygen_test.cc`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-:-KuzbSgwAowQvM6zY3cZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/framework/system_scalar_conversion_doxygen.h, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7091#-KuzbTkqPn20JCqoU-PO:-KuzdZEK_1KwNjDf72Cc:bdhfpta) ([raw file](https://github.com/robotlocomotion/drake/blob/09c52208c39c8fe7f7ba48237b0257c78b551b5b/drake/systems/framework/system_scalar_conversion_doxygen.h#L25)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, do we want/plan to have a copy of the newly added code in this `.h` file in `test/system_scalar_conversion_doxygen_test.cc`?
</blockquote></details>

My thought was that only end-to-end example code was difficult enough to get right to merit a matching unit test.  I think the example System subclasses are relatively simple in comparison, so I wasn't planning on extracting their code into a test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7091)*
<!-- Sent from Reviewable.io -->
